stan/math/opencl/kernels/tri_multiply.hpp

```diff
@@ -0,0 +1,155 @@
+namespace opencl_kernels {
+// \cond
+static const char* tri_multiply_kernel_code = STRINGIFY(
+    // \endcond
+    /**
+     * OpenCL Matrix multiplication of a triangular and rectangular matrix
+     *
+     * @param[in] A the lower/upper triangular matrix in matrix multiplication
+     * @param[in] B the right matrix in matrix multiplication
+     * @param[out] C the output matrix
+     * @param[in] M Number of rows for matrix A
+     * @param[in] N Number of rows for matrix B
+     * @param[in] K Number of cols for matrix A and number of rows for matrix B
+     * @param[in] lower_upper_A the triangularity of A (lower, upper or none)
+     * @param[in] lower_upper_B the triangularity of B (lower, upper or none)
+     */
+    __kernel void tri_multiply(
+        const __global double* A, const __global double* B, __global double* C,
+        const int M, const int N, const int K, unsigned int lower_upper_A,
+        unsigned int lower_upper_B) {
+      // thread index inside the thread_block
+      const int thread_block_row = get_local_id(0);
+      const int thread_block_col = get_local_id(1);
+      // global thread index
+      const int i = THREAD_BLOCK_SIZE * get_group_id(0) + thread_block_row;
+      const int j = THREAD_BLOCK_SIZE * get_group_id(1) + thread_block_col;
+
+      // indexes that determine the last indexes that need to compute
+      // in order to remove the unnecesary multiplications in the special
+      // multiplication of A*A^T
+      const int jMin = THREAD_BLOCK_SIZE * get_group_id(1);
+      const int iMax = THREAD_BLOCK_SIZE * get_group_id(0) + get_local_size(0);
+      // local memory
+      __local double A_local[THREAD_BLOCK_SIZE][THREAD_BLOCK_SIZE];
+      __local double B_local[THREAD_BLOCK_SIZE][THREAD_BLOCK_SIZE];
+
+      double acc[WORK_PER_THREAD];
+      for (int w = 0; w < WORK_PER_THREAD; w++) {
+        acc[w] = 0.0;
+      }
+
+      const int num_tiles = (K + THREAD_BLOCK_SIZE - 1) / THREAD_BLOCK_SIZE;
+
+      const int is_A_lower = lower_upper_A == LOWER;
+      const int is_A_upper = lower_upper_A == UPPER;
+      const int is_B_lower = lower_upper_B == LOWER;
+      const int is_B_upper = lower_upper_B == UPPER;
+      // This kernel is based on the kernel for the regular
+      // matrix multiply where the multiplitcation is split
+      // in tiles.
+      // In cases where a matrix is lower triangular
+      // its not necessary to multiply the elements
+      // over the diagonal, therefore those tiles
+      // in the matrix multiply can be skipped.
+      // With upper triangular matrices we dont need
+      // to multiply the elements under the diagonal,
+      // so thosee tiles can be skipped.
+      // The following code determines the start and
+      // end tile based on triangularity of the input matrices
+      // If no matrices are triangular the starting tile
+      // is 0 and the end tile is num_tiles-1 which
+      // is the same as in \link kernels/multiply.hpp multiply()
```

Comment:
does that mean we can get rid of multiply.hpp's multiply() or outsource that impl to this one?

Reply:
I am going to do another perf test to see if we get any performance regression for the general matrix multiply if we just use this one. My instinct was that the below checks on the copy to `A_local` and `B_local` have a bad impact, hence the 2 kernels. I will double check.

Reply:
Ah okay. That's fine, we can have both.